api/pipeline.go

```diff
@@ -558,130 +324,77 @@ func ValidatePipeline(c *gin.Context) {
-		WithMetadata(m).
-		WithRepo(r).
-		WithUser(u)
-
-	p, err := comp.Parse(config)
-	if err != nil {
-		// nolint: lll // ignore long line length due to error message
-		retErr := fmt.Errorf("unable to parse pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
-
-		util.HandleError(c, http.StatusBadRequest, retErr)
-
+	// validate the yaml configuration
+	if err = comp.Validate(pipeline); err != nil {
+		retErr := fmt.Errorf("unable to validate pipeline configuration for %s: %w", repoName(ctx), err)
+		util.HandleError(ctx, http.StatusBadRequest, retErr)
 		return
 	}
 
-	// create map of templates for easy lookup
-	t := p.Templates.Map()
-
-	// check if the pipeline contains stages
-	if len(p.Stages) > 0 {
-		// inject the templates into the stages
-		p.Stages, p.Secrets, p.Services, p.Environment, err = comp.ExpandStages(p, t)
-		if err != nil {
-			// nolint: lll // ignore long line length due to error message
-			retErr := fmt.Errorf("unable to expand stages in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
-
-			util.HandleError(c, http.StatusBadRequest, retErr)
-
-			return
-		}
-
-		// inject the substituted environment variables into the stages
-		p.Stages, err = comp.SubstituteStages(p.Stages)
-		if err != nil {
-			// nolint: lll // ignore long line length due to error message
-			retErr := fmt.Errorf("unable to substitute stages in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
-
-			util.HandleError(c, http.StatusBadRequest, retErr)
-
-			return
-		}
-	} else {
-		// inject the templates into the steps
-		p.Steps, p.Secrets, p.Services, p.Environment, err = comp.ExpandSteps(p, t)
-		if err != nil {
-			// nolint: lll // ignore long line length due to error message
-			retErr := fmt.Errorf("unable to expand steps in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
-
-			util.HandleError(c, http.StatusBadRequest, retErr)
-
-			return
-		}
+	writeOutput(ctx, pipeline)
+}
 
-		// inject the substituted environment variables into the steps
-		p.Steps, err = comp.SubstituteSteps(p.Steps)
-		if err != nil {
-			// nolint: lll // ignore long line length due to error message
-			retErr := fmt.Errorf("unable to substitute steps in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+// getUnprocessedPipeline retrieves the unprocessed pipeline from a given context.
+func getUnprocessedPipeline(ctx *gin.Context) (*yaml.Build, compiler.Engine, error) {
+	// capture middleware values
+	meta := ctx.MustGet("metadata").(*types.Metadata)
+	repo := repo.Retrieve(ctx)
 
-			util.HandleError(c, http.StatusBadRequest, retErr)
+	// capture query parameters
+	ref := ctx.DefaultQuery("ref", repo.GetBranch())
 
-			return
-		}
+	// send API call to capture the repo owner
+	user, err := database.FromContext(ctx).GetUser(repo.GetUserID())
+	if err != nil {
+		retErr := fmt.Errorf("unable to get owner for %s: %w", repo.GetFullName(), err)
+		return nil, nil, retErr
```

Comment:
I wonder if there is any value in keeping the `retErr` variable here?

Perhaps we condense this to a one liner?

```suggestion
		return nil, nil, fmt.Errorf("unable to get owner for %s: %w", repo.GetFullName(), err)
```

Reply:
I can do a quick pass and clean-up and consolidate a lot of those to one liners. I was following the convention from the rest of the repo and figured it was because the err was being masked and to make it more apparent that the underlying error was being manipulated before being passed back.